{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Classification into groups\n",
    "\n",
    "In the study [Haplotype analyses reveal novel insights into tomato history and domestication including long-distance migrations and latitudinal adaptations](https://www.biorxiv.org/content/10.1101/2021.06.18.448912v2) tomato accessions were genotyped by Whole Genome Sequencing.\n",
    "\n",
    "We have the genotypes in the file \"tomato_genotypes.csv.gz\" and we want to classify some of them into groups.\n",
    "The accessions to classify and its classification is provided in the Python dictionary named \"classification\".\n",
    "Our task is to replace the 'none' classification set for every accession with a suitable classification.\n",
    "\n",
    "We will classify the accessions by doing a Principal Component Analysis with the genotypes. Also, we have passport data in the file \"suppl_table_2_passport_and_morphological_data.csv.gz\". For example, we have the passport \"Taxon\" and \"Country\". So, we want to classify the accessions into groups by taking into account the Principal Component Analysis result and this passport data.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import plotly.express as px\n",
    "import pca\n",
    "\n",
    "classification = '''\n",
    "bgv005895 none\n",
    "bgv005912 none\n",
    "bgv006234 none\n",
    "bgv006235 none\n",
    "bgv006327 none\n",
    "bgv006336 none\n",
    "bgv006347 none\n",
    "bgv006353 none\n",
    "bgv006370 none\n",
    "bgv006454 none\n",
    "bgv006457 none\n",
    "bgv006767 none\n",
    "bgv006775 none\n",
    "bgv006825 none\n",
    "bgv006852 none\n",
    "bgv006859 none\n",
    "bgv006867 none\n",
    "bgv006904 none\n",
    "bgv006910 none\n",
    "bgv006931 none\n",
    "bgv006934 none\n",
    "bgv007023 none\n",
    "bgv007111 none\n",
    "bgv007149 none\n",
    "bgv007151 none\n",
    "bgv007152 none\n",
    "bgv007158 none\n",
    "bgv007161 none\n",
    "bgv007169 none\n",
    "bgv007181 none\n",
    "bgv007194 none\n",
    "bgv007198 none\n",
    "bgv007854 none\n",
    "bgv007857 none\n",
    "bgv007860 none\n",
    "bgv007862 none\n",
    "bgv007863 none\n",
    "bgv007864 none\n",
    "bgv007865 none\n",
    "bgv007867 none\n",
    "bgv007870 none\n",
    "bgv007871 none\n",
    "bgv007872 none\n",
    "bgv007875 none\n",
    "bgv007876 none\n",
    "bgv007878 none\n",
    "bgv007901 none\n",
    "bgv007902 none\n",
    "bgv007908 none\n",
    "bgv007909 none\n",
    "bgv007910 none\n",
    "bgv007911 none\n",
    "bgv007918 none\n",
    "bgv007921 none\n",
    "bgv007927 none\n",
    "bgv007931 none\n",
    "bgv007933 none\n",
    "bgv007934 none\n",
    "bgv007936 none\n",
    "bgv007981 none\n",
    "bgv007989 none\n",
    "bgv008036 none\n",
    "bgv008037 none\n",
    "bgv008041 none\n",
    "bgv008051 none\n",
    "bgv008058 none\n",
    "bgv008061 none\n",
    "bgv008065 none\n",
    "bgv008067 none\n",
    "bgv008070 none\n",
    "bgv008077 none\n",
    "bgv008095 none\n",
    "bgv008098 none\n",
    "bgv008100 none\n",
    "bgv008106 none\n",
    "bgv008108 none\n",
    "bgv008218 none\n",
    "bgv008219 none\n",
    "bgv008223 none\n",
    "bgv008224 none\n",
    "bgv008225 none\n",
    "bgv008345 none\n",
    "bgv008348 none\n",
    "bgv012615 none\n",
    "bgv012625 none\n",
    "bgv012626 none\n",
    "bgv012627 none\n",
    "bgv012640 none\n",
    "bgv013134 none\n",
    "bgv013161 none\n",
    "bgv013175 none\n",
    "bgv013945 none\n",
    "bgv015726 none\n",
    "bgv015734 none\n",
    "la1712 none\n",
    "'''\n",
    "\n",
    "classification = dict(line.strip().split() for line in classification.splitlines() if line.strip())\n",
    "\n",
    "taxon_column_name = 'Taxon'\n",
    "country_column_name = 'Country'\n",
    "clasification_column_name = 'Classification'\n",
    "\n",
    "def plot_projections_by_taxon_and_country(pca_result, passport_data, color_column=None):\n",
    "\n",
    "    projections = pca_result['projections']\n",
    "    dframe_to_plot = projections.join(passport_data)\n",
    "\n",
    "    fig = px.scatter(dframe_to_plot, x='princomp_1', y='princomp_2', color=color_column,\n",
    "                     hover_data=[dframe_to_plot.index, \"Taxon\", \"Country\", \"Classification\"])\n",
    "    fig.show()\n",
    "\n",
    "\n",
    "genotypes = pandas.read_csv('tomato_genotypes.csv.gz', index_col=0, header=0)\n",
    "accession_passport_data = pandas.read_csv('tomato_passport_and_morphological_data.csv.gz',\n",
    "                                          skiprows=22,\n",
    "                                          index_col=0)\n",
    "accession_passport_data.index = list(map(lambda x: x.lower(), accession_passport_data.index))\n",
    "classifications = [classification.get(accession, 'nd') for accession in accession_passport_data.index]\n",
    "accession_data = accession_passport_data\n",
    "accession_data[clasification_column_name] = classifications\n",
    "\n",
    "accessions_to_use = sorted(classification.keys())\n",
    "genotypes = genotypes.loc[:, accessions_to_use]\n",
    "\n",
    "pca_result = pca.do_pca_from_012_genotypes(genotypes)\n",
    "\n",
    "color_plot = clasification_column_name\n",
    "plot_projections_by_taxon_and_country(pca_result, accession_data, color_column=color_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Classify the individuals into groups\n",
    "\n",
    "The previous code will run a PCA analysis based on the genotypes, and will plot the results according to the classification given.\n",
    "Your objective is to run the PCA and to create groups/populations. Once you decide on a classification for any individual modify the classification in the text. At the begining all individuals are classified as belonging to the class \"none\".\n",
    "\n",
    "Hovering above any dot will show its taxonomic classification, its country of origin, and its current classification. Also, you can choose one of those traits to be use for the color of the dots in the plot. The information used is defined in the variable color_plot. By default the classification is used for the color, but you can change this by changing the line:\n",
    "\n",
    "```\n",
    "color_plot = clasification_column_name\n",
    "```\n",
    "\n",
    "With:\n",
    "\n",
    "```\n",
    "color_plot = taxon_column_name\n",
    "```\n",
    "\n",
    "Or with:\n",
    "\n",
    "```\n",
    "color_plot = country_column_name\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "91333ed49fc9d0bb54b7c884f80045a87accdf6a1a88fe176e471380fcbbe7ed"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit ('pyenv': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
