{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Classification into groups\n",
    "\n",
    "In the study [Haplotype analyses reveal novel insights into tomato history and domestication including long-distance migrations and latitudinal adaptations](https://www.biorxiv.org/content/10.1101/2021.06.18.448912v2) tomato accessions were genotyped by Whole Genome Sequencing.\n",
    "\n",
    "We have the genotypes in the file \"tomato_genotypes.csv.gz\" and we want to classify some of them into groups.\n",
    "The accessions to classify and its classification is provided in the Python dictionary named \"classification\".\n",
    "Our task is to replace the 'none' classification set for every accession with a suitable classification.\n",
    "\n",
    "We will classify the accessions by doing a Principal Component Analysis with the genotypes. Also, we have passport data in the file \"suppl_table_2_passport_and_morphological_data.csv.gz\". For example, we have the passport \"Taxon\" and \"Country\". So, we want to classify the accessions into groups by taking into account the Principal Component Analysis result and this passport data.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import plotly.express as px\n",
    "import pca\n",
    "\n",
    "\n",
    "classification = {\n",
    "    'bgv005895': 'none', 'bgv005912': 'none', 'bgv006234': 'none',\n",
    "    'bgv006235': 'none', 'bgv006327': 'none', 'bgv006336': 'none',\n",
    "    'bgv006347': 'none', 'bgv006353': 'none', 'bgv006370': 'none',\n",
    "    'bgv006454': 'none', 'bgv006457': 'none', 'bgv006767': 'none',\n",
    "    'bgv006775': 'none', 'bgv006825': 'none', 'bgv006852': 'none',\n",
    "    'bgv006859': 'none', 'bgv006867': 'none', 'bgv006904': 'none',\n",
    "    'bgv006910': 'none', 'bgv006931': 'none', 'bgv006934': 'none',\n",
    "    'bgv007023': 'none', 'bgv007111': 'none', 'bgv007149': 'none',\n",
    "    'bgv007151': 'none', 'bgv007152': 'none', 'bgv007158': 'none',\n",
    "    'bgv007161': 'none', 'bgv007169': 'none', 'bgv007181': 'none',\n",
    "    'bgv007194': 'none', 'bgv007198': 'none', 'bgv007854': 'none',\n",
    "    'bgv007857': 'none', 'bgv007860': 'none', 'bgv007862': 'none',\n",
    "    'bgv007863': 'none', 'bgv007864': 'none', 'bgv007865': 'none',\n",
    "    'bgv007867': 'none', 'bgv007870': 'none', 'bgv007871': 'none',\n",
    "    'bgv007872': 'none', 'bgv007875': 'none', 'bgv007876': 'none',\n",
    "    'bgv007878': 'none', 'bgv007901': 'none', 'bgv007902': 'none',\n",
    "    'bgv007908': 'none', 'bgv007909': 'none', 'bgv007910': 'none',\n",
    "    'bgv007911': 'none', 'bgv007918': 'none', 'bgv007921': 'none',\n",
    "    'bgv007927': 'none', 'bgv007931': 'none', 'bgv007933': 'none',\n",
    "    'bgv007934': 'none', 'bgv007936': 'none', 'bgv007981': 'none',\n",
    "    'bgv007989': 'none', 'bgv008036': 'none', 'bgv008037': 'none',\n",
    "    'bgv008041': 'none', 'bgv008051': 'none', 'bgv008058': 'none',\n",
    "    'bgv008061': 'none', 'bgv008065': 'none', 'bgv008067': 'none',\n",
    "    'bgv008070': 'none', 'bgv008077': 'none', 'bgv008095': 'none',\n",
    "    'bgv008098': 'none', 'bgv008100': 'none', 'bgv008106': 'none',\n",
    "    'bgv008108': 'none', 'bgv008218': 'none', 'bgv008219': 'none',\n",
    "    'bgv008223': 'none', 'bgv008224': 'none', 'bgv008225': 'none',\n",
    "    'bgv008345': 'none', 'bgv008348': 'none', 'bgv012615': 'none',\n",
    "    'bgv012625': 'none', 'bgv012626': 'none', 'bgv012627': 'none',\n",
    "    'bgv012640': 'none', 'bgv013134': 'none', 'bgv013161': 'none',\n",
    "    'bgv013175': 'none', 'bgv013945': 'none', 'bgv015726': 'none',\n",
    "    'bgv015734': 'none', 'la1712': 'none'}\n",
    "\n",
    "\n",
    "taxon_column_name = 'Taxon'\n",
    "country_column_name = 'Country'\n",
    "clasification_column_name = 'Classification'\n",
    "\n",
    "def plot_projections_by_taxon_and_country(pca_result, passport_data, color_column=None):\n",
    "\n",
    "    projections = data=pca_result['projections']\n",
    "    dframe_to_plot = projections.join(passport_data)\n",
    "    fig = px.scatter(dframe_to_plot, x='princomp_1', y='princomp_2', color=color_column,\n",
    "                     hover_data=[dframe_to_plot.index, \"Taxon\", \"Country\", \"Classification\"])\n",
    "    fig.show()\n",
    "\n",
    "\n",
    "genotypes = pandas.read_csv('tomato_genotypes.csv.gz')\n",
    "accession_passport_data = pandas.read_csv('tomato_passport_and_morphological_data.csv.gz',\n",
    "                                          skiprows=22,\n",
    "                                          index_col=0)\n",
    "accession_passport_data.index = list(map(lambda x: x.lower(), accession_passport_data.index))\n",
    "classifications = [classification.get(accession, 'nd') for accession in accession_passport_data.index]\n",
    "accession_data = accession_passport_data\n",
    "accession_data[clasification_column_name] = classifications\n",
    "\n",
    "accessions_to_use = sorted(classification.keys())\n",
    "genotypes = genotypes.loc[:, accessions_to_use]\n",
    "\n",
    "pca_result = pca.do_pca_from_012_genotypes(genotypes)\n",
    "\n",
    "plot_projections_by_taxon_and_country(pca_result, accession_data)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "91333ed49fc9d0bb54b7c884f80045a87accdf6a1a88fe176e471380fcbbe7ed"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit ('pyenv': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
