{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Classification into groups\n",
    "\n",
    "In the study [Haplotype analyses reveal novel insights into tomato history and domestication including long-distance migrations and latitudinal adaptations](https://www.biorxiv.org/content/10.1101/2021.06.18.448912v2) tomato accessions were genotyped by Whole Genome Sequencing.\n",
    "\n",
    "We have the genotypes in the file \"tomato_genotypes.csv.gz\" and we want to classify some of them into groups.\n",
    "The accessions to classify and its classification is provided in the Python dictionary named \"classification\".\n",
    "Our task is to replace the 'none' classification set for every accession with a suitable classification.\n",
    "\n",
    "We will classify the accessions by doing a Principal Component Analysis with the genotypes. Also, we have passport data in the file \"suppl_table_2_passport_and_morphological_data.csv.gz\". For example, we have the passport \"Taxon\" and \"Country\". So, we want to classify the accessions into groups by taking into account the Principal Component Analysis result and this passport data.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       bgv014519  bgv015730  tegucigalpa  pi378994  pi487625  la2697  \\\n",
      "0              0          0            0         2         0       0   \n",
      "1              0          0            0         0         0       2   \n",
      "2              0          0            0         2         0       0   \n",
      "3              0          0            0         2         0       0   \n",
      "4              0          0            0         2         0       0   \n",
      "...          ...        ...          ...       ...       ...     ...   \n",
      "33785          0          0            0         0         0       0   \n",
      "33786          0          0            0         0         0       0   \n",
      "33787          0          0            0         0         0       0   \n",
      "33788          0          0            0         0         0       0   \n",
      "33789          0          0            0         0         0       0   \n",
      "\n",
      "       bgv015726  la2309  ts-588  ts-138  ...  bgv013161  bgv007910  \\\n",
      "0              0       0      -1       0  ...          0          0   \n",
      "1              0       0       0       0  ...          0          0   \n",
      "2              0       0       0       0  ...          0          0   \n",
      "3              0       0       0       0  ...          0          0   \n",
      "4              0       0       0       0  ...          0          0   \n",
      "...          ...     ...     ...     ...  ...        ...        ...   \n",
      "33785          0       0       0       0  ...          0          0   \n",
      "33786          0       0      -1       0  ...          0          0   \n",
      "33787          0       0       0       0  ...          0          0   \n",
      "33788          0       0       0       0  ...          0          0   \n",
      "33789          0       0       0       0  ...          0          0   \n",
      "\n",
      "       bgv007909  bgv012625  bgv015382  ts-132  ts-170  bgv014518  ts-518  \\\n",
      "0              0          0          2       0      -1          0       0   \n",
      "1              0          2         -1       0      -1          0       0   \n",
      "2              0          0         -1       0       0          0       0   \n",
      "3              0          2         -1       0      -1          0       0   \n",
      "4              0          2          2       0       0          0       0   \n",
      "...          ...        ...        ...     ...     ...        ...     ...   \n",
      "33785          0          0          2       0       0          0       0   \n",
      "33786          0          0          0       0      -1          0       0   \n",
      "33787          0          0          0       0       0          0       0   \n",
      "33788          0          0         -1       0       0          0       0   \n",
      "33789          0          0          0       0      -1          0       0   \n",
      "\n",
      "       catie-11106-1  \n",
      "0                  0  \n",
      "1                  0  \n",
      "2                  0  \n",
      "3                  0  \n",
      "4                  0  \n",
      "...              ...  \n",
      "33785              0  \n",
      "33786              0  \n",
      "33787              0  \n",
      "33788              0  \n",
      "33789              0  \n",
      "\n",
      "[33790 rows x 598 columns]\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Shape of passed values is (95, 10), indices imply (33790, 10)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m/home/jose/analyses/population_genetics_simulations/src/tomato_pop_definition_practice.ipynb Cell 2'\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/jose/analyses/population_genetics_simulations/src/tomato_pop_definition_practice.ipynb#ch0000001?line=64'>65</a>\u001b[0m accessions_to_use \u001b[39m=\u001b[39m \u001b[39msorted\u001b[39m(classification\u001b[39m.\u001b[39mkeys())\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/jose/analyses/population_genetics_simulations/src/tomato_pop_definition_practice.ipynb#ch0000001?line=65'>66</a>\u001b[0m genotypes \u001b[39m=\u001b[39m genotypes\u001b[39m.\u001b[39mloc[:, accessions_to_use]\n\u001b[0;32m---> <a href='vscode-notebook-cell:/home/jose/analyses/population_genetics_simulations/src/tomato_pop_definition_practice.ipynb#ch0000001?line=67'>68</a>\u001b[0m pca_result \u001b[39m=\u001b[39m pca\u001b[39m.\u001b[39;49mdo_pca_from_012_genotypes(genotypes)\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/jose/analyses/population_genetics_simulations/src/tomato_pop_definition_practice.ipynb#ch0000001?line=69'>70</a>\u001b[0m plot_projections_by_taxon_and_country(pca_result, accession_data)\n",
      "File \u001b[0;32m~/analyses/population_genetics_simulations/src/pca.py:24\u001b[0m, in \u001b[0;36mdo_pca_from_012_genotypes\u001b[0;34m(genotypes, n_components, scaler, use_randomized_pca_optimization)\u001b[0m\n\u001b[1;32m     <a href='file:///~/analyses/population_genetics_simulations/src/pca.py?line=20'>21</a>\u001b[0m explained_variances \u001b[39m=\u001b[39m pandas\u001b[39m.\u001b[39mSeries(explained_variances, index\u001b[39m=\u001b[39mprincom_names)\n\u001b[1;32m     <a href='file:///~/analyses/population_genetics_simulations/src/pca.py?line=21'>22</a>\u001b[0m explained_variances \u001b[39m=\u001b[39m explained_variances \u001b[39m*\u001b[39m \u001b[39m100.0\u001b[39m\n\u001b[0;32m---> <a href='file:///~/analyses/population_genetics_simulations/src/pca.py?line=23'>24</a>\u001b[0m projections \u001b[39m=\u001b[39m pandas\u001b[39m.\u001b[39;49mDataFrame(\n\u001b[1;32m     <a href='file:///~/analyses/population_genetics_simulations/src/pca.py?line=24'>25</a>\u001b[0m     projections, columns\u001b[39m=\u001b[39;49mprincom_names, index\u001b[39m=\u001b[39;49mgenotypes\u001b[39m.\u001b[39;49mindex\n\u001b[1;32m     <a href='file:///~/analyses/population_genetics_simulations/src/pca.py?line=25'>26</a>\u001b[0m )\n\u001b[1;32m     <a href='file:///~/analyses/population_genetics_simulations/src/pca.py?line=27'>28</a>\u001b[0m \u001b[39mreturn\u001b[39;00m {\u001b[39m\"\u001b[39m\u001b[39mprojections\u001b[39m\u001b[39m\"\u001b[39m: projections, \u001b[39m\"\u001b[39m\u001b[39mexplained_variances\u001b[39m\u001b[39m\"\u001b[39m: explained_variances}\n",
      "File \u001b[0;32m~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py:695\u001b[0m, in \u001b[0;36mDataFrame.__init__\u001b[0;34m(self, data, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=684'>685</a>\u001b[0m         mgr \u001b[39m=\u001b[39m dict_to_mgr(\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=685'>686</a>\u001b[0m             \u001b[39m# error: Item \"ndarray\" of \"Union[ndarray, Series, Index]\" has no\u001b[39;00m\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=686'>687</a>\u001b[0m             \u001b[39m# attribute \"name\"\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=691'>692</a>\u001b[0m             typ\u001b[39m=\u001b[39mmanager,\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=692'>693</a>\u001b[0m         )\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=693'>694</a>\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=694'>695</a>\u001b[0m         mgr \u001b[39m=\u001b[39m ndarray_to_mgr(\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=695'>696</a>\u001b[0m             data,\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=696'>697</a>\u001b[0m             index,\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=697'>698</a>\u001b[0m             columns,\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=698'>699</a>\u001b[0m             dtype\u001b[39m=\u001b[39;49mdtype,\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=699'>700</a>\u001b[0m             copy\u001b[39m=\u001b[39;49mcopy,\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=700'>701</a>\u001b[0m             typ\u001b[39m=\u001b[39;49mmanager,\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=701'>702</a>\u001b[0m         )\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=703'>704</a>\u001b[0m \u001b[39m# For data is list-like, or Iterable (will consume into list)\u001b[39;00m\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/frame.py?line=704'>705</a>\u001b[0m \u001b[39melif\u001b[39;00m is_list_like(data):\n",
      "File \u001b[0;32m~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/internals/construction.py:351\u001b[0m, in \u001b[0;36mndarray_to_mgr\u001b[0;34m(values, index, columns, dtype, copy, typ)\u001b[0m\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/internals/construction.py?line=345'>346</a>\u001b[0m \u001b[39m# _prep_ndarray ensures that values.ndim == 2 at this point\u001b[39;00m\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/internals/construction.py?line=346'>347</a>\u001b[0m index, columns \u001b[39m=\u001b[39m _get_axes(\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/internals/construction.py?line=347'>348</a>\u001b[0m     values\u001b[39m.\u001b[39mshape[\u001b[39m0\u001b[39m], values\u001b[39m.\u001b[39mshape[\u001b[39m1\u001b[39m], index\u001b[39m=\u001b[39mindex, columns\u001b[39m=\u001b[39mcolumns\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/internals/construction.py?line=348'>349</a>\u001b[0m )\n\u001b[0;32m--> <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/internals/construction.py?line=350'>351</a>\u001b[0m _check_values_indices_shape_match(values, index, columns)\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/internals/construction.py?line=352'>353</a>\u001b[0m \u001b[39mif\u001b[39;00m typ \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39marray\u001b[39m\u001b[39m\"\u001b[39m:\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/internals/construction.py?line=354'>355</a>\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39missubclass\u001b[39m(values\u001b[39m.\u001b[39mdtype\u001b[39m.\u001b[39mtype, \u001b[39mstr\u001b[39m):\n",
      "File \u001b[0;32m~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/internals/construction.py:422\u001b[0m, in \u001b[0;36m_check_values_indices_shape_match\u001b[0;34m(values, index, columns)\u001b[0m\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/internals/construction.py?line=419'>420</a>\u001b[0m passed \u001b[39m=\u001b[39m values\u001b[39m.\u001b[39mshape\n\u001b[1;32m    <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/internals/construction.py?line=420'>421</a>\u001b[0m implied \u001b[39m=\u001b[39m (\u001b[39mlen\u001b[39m(index), \u001b[39mlen\u001b[39m(columns))\n\u001b[0;32m--> <a href='file:///~/analyses/population_genetics_simulations/pyenv/lib/python3.9/site-packages/pandas/core/internals/construction.py?line=421'>422</a>\u001b[0m \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mShape of passed values is \u001b[39m\u001b[39m{\u001b[39;00mpassed\u001b[39m}\u001b[39;00m\u001b[39m, indices imply \u001b[39m\u001b[39m{\u001b[39;00mimplied\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n",
      "\u001b[0;31mValueError\u001b[0m: Shape of passed values is (95, 10), indices imply (33790, 10)"
     ]
    }
   ],
   "source": [
    "import pandas\n",
    "import plotly.express as px\n",
    "import pca\n",
    "\n",
    "\n",
    "classification = {\n",
    "    'bgv005895': 'none', 'bgv005912': 'none', 'bgv006234': 'none',\n",
    "    'bgv006235': 'none', 'bgv006327': 'none', 'bgv006336': 'none',\n",
    "    'bgv006347': 'none', 'bgv006353': 'none', 'bgv006370': 'none',\n",
    "    'bgv006454': 'none', 'bgv006457': 'none', 'bgv006767': 'none',\n",
    "    'bgv006775': 'none', 'bgv006825': 'none', 'bgv006852': 'none',\n",
    "    'bgv006859': 'none', 'bgv006867': 'none', 'bgv006904': 'none',\n",
    "    'bgv006910': 'none', 'bgv006931': 'none', 'bgv006934': 'none',\n",
    "    'bgv007023': 'none', 'bgv007111': 'none', 'bgv007149': 'none',\n",
    "    'bgv007151': 'none', 'bgv007152': 'none', 'bgv007158': 'none',\n",
    "    'bgv007161': 'none', 'bgv007169': 'none', 'bgv007181': 'none',\n",
    "    'bgv007194': 'none', 'bgv007198': 'none', 'bgv007854': 'none',\n",
    "    'bgv007857': 'none', 'bgv007860': 'none', 'bgv007862': 'none',\n",
    "    'bgv007863': 'none', 'bgv007864': 'none', 'bgv007865': 'none',\n",
    "    'bgv007867': 'none', 'bgv007870': 'none', 'bgv007871': 'none',\n",
    "    'bgv007872': 'none', 'bgv007875': 'none', 'bgv007876': 'none',\n",
    "    'bgv007878': 'none', 'bgv007901': 'none', 'bgv007902': 'none',\n",
    "    'bgv007908': 'none', 'bgv007909': 'none', 'bgv007910': 'none',\n",
    "    'bgv007911': 'none', 'bgv007918': 'none', 'bgv007921': 'none',\n",
    "    'bgv007927': 'none', 'bgv007931': 'none', 'bgv007933': 'none',\n",
    "    'bgv007934': 'none', 'bgv007936': 'none', 'bgv007981': 'none',\n",
    "    'bgv007989': 'none', 'bgv008036': 'none', 'bgv008037': 'none',\n",
    "    'bgv008041': 'none', 'bgv008051': 'none', 'bgv008058': 'none',\n",
    "    'bgv008061': 'none', 'bgv008065': 'none', 'bgv008067': 'none',\n",
    "    'bgv008070': 'none', 'bgv008077': 'none', 'bgv008095': 'none',\n",
    "    'bgv008098': 'none', 'bgv008100': 'none', 'bgv008106': 'none',\n",
    "    'bgv008108': 'none', 'bgv008218': 'none', 'bgv008219': 'none',\n",
    "    'bgv008223': 'none', 'bgv008224': 'none', 'bgv008225': 'none',\n",
    "    'bgv008345': 'none', 'bgv008348': 'none', 'bgv012615': 'none',\n",
    "    'bgv012625': 'none', 'bgv012626': 'none', 'bgv012627': 'none',\n",
    "    'bgv012640': 'none', 'bgv013134': 'none', 'bgv013161': 'none',\n",
    "    'bgv013175': 'none', 'bgv013945': 'none', 'bgv015726': 'none',\n",
    "    'bgv015734': 'none', 'la1712': 'none'}\n",
    "\n",
    "\n",
    "taxon_column_name = 'Taxon'\n",
    "country_column_name = 'Country'\n",
    "clasification_column_name = 'Classification'\n",
    "\n",
    "def plot_projections_by_taxon_and_country(pca_result, passport_data, color_column=None):\n",
    "\n",
    "    projections = pca_result['projections']\n",
    "    dframe_to_plot = projections.join(passport_data)\n",
    "\n",
    "    fig = px.scatter(dframe_to_plot, x='princomp_1', y='princomp_2', color=color_column,\n",
    "                     hover_data=[dframe_to_plot.index, \"Taxon\", \"Country\", \"Classification\"])\n",
    "    fig.show()\n",
    "\n",
    "\n",
    "genotypes = pandas.read_csv('tomato_genotypes.csv.gz', index_col=0, header=0)\n",
    "accession_passport_data = pandas.read_csv('tomato_passport_and_morphological_data.csv.gz',\n",
    "                                          skiprows=22,\n",
    "                                          index_col=0)\n",
    "accession_passport_data.index = list(map(lambda x: x.lower(), accession_passport_data.index))\n",
    "classifications = [classification.get(accession, 'nd') for accession in accession_passport_data.index]\n",
    "accession_data = accession_passport_data\n",
    "accession_data[clasification_column_name] = classifications\n",
    "\n",
    "accessions_to_use = sorted(classification.keys())\n",
    "genotypes = genotypes.loc[:, accessions_to_use]\n",
    "\n",
    "pca_result = pca.do_pca_from_012_genotypes(genotypes)\n",
    "\n",
    "plot_projections_by_taxon_and_country(pca_result, accession_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "91333ed49fc9d0bb54b7c884f80045a87accdf6a1a88fe176e471380fcbbe7ed"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit ('pyenv': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
